lib/shipit/stack_commands.rb

```diff
@@ -15,19 +15,25 @@ def env
 
     def fetch
       create_directories
-      if Dir.exist?(@stack.git_path)
+      if valid_git_repository?(@stack.git_path)
         git('fetch', 'origin', '--tags', @stack.branch, env: env, chdir: @stack.git_path)
       else
+        @stack.clear_git_cache!(with_lock: false)
```

Comment:
If the stack's git cache repo is present, but corrupt, the subsequent call to `git_clone` will exit with status 128. The previous iteration looks like it assumed that if we entered this branch that the git was absent anyway. I _think_ this means that ensuring the precondition of an empty git cache here would be OK, or even desirable. 

The previous version of the `Stack#clear_git_cache!` method assumed that it is running in a context which has NOT already obtained the file-lock on the git cache. So far as I was able to determine this `#fetch` method is only called in a context where the outer scope has already acquired the git cache filesystem lock - while [executing a task / deployment](https://github.com/Shopify/shipit-engine/blob/master/app/models/shipit/task_execution_strategy/default.rb#L70-L76) and when [fetching fetching a commit's ref into a temporary working directory](https://github.com/Shopify/shipit-engine/blob/0edf50782d5b8d2b6e1976f96d35937a84c4c1b1/lib/shipit/stack_commands.rb#L54-L58). Since the outer scopes which call the `#fetch` method have already acquired the git cache file lock, this needs a way to delete the git cache without trying to reacquire the git cache lock a 2nd time - since this would cause a flock error.

I chose to add the ability to tell `clear_git_cache!` to skip trying to reacquire the git cache lock because we already have it. This prevents duplication of the logic to delete the git cache, but felt a bit goofy. Would love opinions and alternate points of view on this.

Reply:
Thanks @casperisfine ,  I was - slowly - working on something similar. I'll rebase and remove the `with_lock` bits. Appreciate the help.